core/domain/improvements_services.py

```diff
@@ -0,0 +1,193 @@
+            feconf.MAX_TASK_MODELS_PER_FETCH, start_cursor=cursor)
+        for task_model in results:
+            yield get_task_entry_from_model(task_model)
+
+
+def get_task_entry_from_model(task_entry_model):
+    """Returns a domain object for the corresponding task entry model.
+
+    Args:
+        task_entry_model: improvements_models.TaskEntryModel.
+
+    Returns:
+        improvements_domain.TaskEntry.
+    """
+    return improvements_domain.TaskEntry(
+        task_entry_model.entity_type, task_entry_model.entity_id,
+        task_entry_model.entity_version, task_entry_model.task_type,
+        task_entry_model.target_type, task_entry_model.target_id,
+        task_entry_model.issue_description, task_entry_model.status,
+        task_entry_model.resolver_id, task_entry_model.resolved_on)
+
+
+def fetch_exploration_tasks(exploration):
+    """Returns a tuple encoding all of the tasks corresponding to the provided
+    exploration.
+
+    Args:
+        exploration: exp_domain.Exploration.
+
+    Returns:
+        tuple. Contains the following 2 items:
+            open_tasks: list(improvements_domain.TaskEntry). The list of the
+                tasks which are open.
+            resolved_task_types_by_state_name: dict(str: list(str)). Maps state
+                names to the list of resolved task types. Absent keys imply that
+                the state has no resolved tasks.
+    """
+    composite_entity_id = (
+        improvements_models.TaskEntryModel.generate_composite_entity_id(
+            improvements_models.TASK_ENTITY_TYPE_EXPLORATION,
+            exploration.id, exploration.version))
+    tasks_grouped_by_status = itertools.groupby(
+        _yield_all_tasks_ordered_by_status(composite_entity_id),
+        operator.attrgetter('status'))
+
+    open_tasks = []
+    resolved_task_types_by_state_name = collections.defaultdict(list)
+    for status_group, tasks in tasks_grouped_by_status:
+        if status_group == improvements_models.TASK_STATUS_OPEN:
+            open_tasks.extend(tasks)
+        elif status_group == improvements_models.TASK_STATUS_RESOLVED:
+            for t in tasks:
+                resolved_task_types_by_state_name[t.target_id].append(
+                    t.task_type)
+    return (open_tasks, dict(resolved_task_types_by_state_name))
+
+
+def fetch_exploration_task_history_page(exploration, cursor=None):
+    """Fetches a page of tasks from the provided entity's history of tasks.
+
+    Args:
+        exploration: exp_domain.Exploration.
+        cursor: datastore_query.Cursor or None. Starting point for search. When
+            None, the starting point is the very beginning of the history
+            results (i.e., starting from the newest task entry).
+
+    Returns:
+        tuple. Contains the following 3 items:
+            results: list(improvements_domain.TaskEntry). The query results.
+            cursor: datastore_query.Cursor or None. a query cursor pointing to
```

Comment:
Wouldn't this be passed in from the frontend? I thought you'd need a urlsafe string cursor here, similar to what we do elsewhere in the codebase (see e.g. core/domain/feedback_services.py).

(Even if not, might be more prudent to do the string version since (a) it is consistent with what we do elsewhere, (b) you may pass this to the frontend later.)

Reply:
Done.